### PR TITLE
add watchdog

### PR DESCRIPTION
**Pull request by hervenicol**
Towards https://github.com/giantswarm/giantswarm/issues/25400

This PR:

- adds a watchdog to reset agent if too much data is enqueued, to reduce stress on remote prometheus

### Checklist

- [x] Update changelog in CHANGELOG.md.
- [ ] Test on Workload cluster.

### CODE REVIEW

**Comment by hervenicol** [comment]
Tested on `gauss`:
- killed prometheus a few times
- when prometheus started, prometheus-agent had around 700 "desired" chards
- liveness failed, prometheus-agent restarted
- after the restart, it had dropped all queued metrics, so was at 0 "desired" chards.

**Review by QuantumEnigmaa — APPROVED** [review]

**Comment by QuentinBisson** [comment]
LGTM but did you check if it's not crashing too often?

**Comment by hervenicol** [comment]
> LGTM but did you check if it's not crashing too often?

From my tests on `gauss` it looked good.
If it breaks too much, we should tune it (maybe tuning `overload_tolerance` in the script).

Brings the question of the process when tuning it.
Up to now, it meant a new release, that would be deployed all around.
Now, with o11y bundle, GS releases, and so on, maybe we should plan for managing as much stuff as possible from the config, so we can tune without re-releasing?

**Comment by hervenicol** [comment]
OK @hervenicol , here are a few thing that you should fix before merging:
- the subchart's livenessProbe relies on a script mounted by the top-level chart as configmap. That makes it dependent on the top-level chart, which we don't want.
- we should have a way to enable/disable the watchdog from `values`
- a way to tune the watchdog from `values` would be nice too.

**Review by TheoBrigitte — CHANGES_REQUESTED** [review]
Nice job, but got some remarks.

File: helm/prometheus-agent/charts/prometheus-agent/files/watchdog.sh
Comment (hervenicol) [reply helm/prometheus-agent/charts/prometheus-agent/files/watchdog.sh:25]: hmm, nice catch.

Short answer: I don't know.

Long answer:
I took note of the default (httpGet) liveness probe before changing anything.
Then I switched to "ListenLocal" mode, which used a "command" probe.
And I copied the command from the "command" probe to put it in my script.

But now I looked at the code:
* [definition of the probes](https://github.com/prometheus-operator/prometheus-operator/blob/ef1958adf24c8c793e42c7a5f75f464a7b8c50d8/pkg/prometheus/statefulset.go#L616)
* [comment about which path to use](https://github.com/prometheus-operator/prometheus-operator/blob/ef1958adf24c8c793e42c7a5f75f464a7b8c50d8/pkg/prometheus/statefulset.go#L649)

...it looks clear I should stick to `/-/healthy`.
I'll fix it.

File: helm/prometheus-agent/charts/prometheus-agent/files/watchdog.sh
Comment (hervenicol) [reply helm/prometheus-agent/charts/prometheus-agent/files/watchdog.sh:35]: :astonished: sure, and my `sed` as well. To fit the metric name.

File: helm/prometheus-agent/charts/prometheus-agent/files/watchdog.sh
Comment (hervenicol) [reply helm/prometheus-agent/charts/prometheus-agent/files/watchdog.sh:25]: ...about the number of tries: it's also kept in the probe definition.
I added a retry to `wget` because, as we run it 3 times in the script, chances that it fails are higher.

File: helm/prometheus-agent/charts/prometheus-agent/files/watchdog.sh
Comment (hervenicol) [reply helm/prometheus-agent/charts/prometheus-agent/files/watchdog.sh:38]: I assume if `/metrics` does not return valid data, there's something wrong with our prometheus-agent.

File: helm/prometheus-agent/values.yaml
Comment (hervenicol) [reply helm/prometheus-agent/values.yaml:16]: oh, that's a nice one: adding a template to the parent chart required to interpret _helpers, which required this key to exist.
As I since moved the template to the subchart, I don't need this one anymore.

**Comment by hervenicol** [comment]
Tested overloading values on `gauss` looks all good.

File: helm/prometheus-agent/charts/prometheus-agent/files/watchdog.sh
Comment (hervenicol) [reply helm/prometheus-agent/charts/prometheus-agent/files/watchdog.sh]: Removed the misleading comments.

File: helm/prometheus-agent/charts/prometheus-agent/files/watchdog.sh
Comment (TheoBrigitte) [reply helm/prometheus-agent/charts/prometheus-agent/files/watchdog.sh:34]: Agreed that we prefer to have the pod being killed repeatedly when this script fails (maybe the metrics name changed, maybe something changes in the agent), rather than have it fails silently.
So someone will actually go and check this instead of loosing the watchdog functionality without knowing.

**Review by TheoBrigitte — APPROVED** [review]

**Pull request merged**
